Migrate to TS, `bottlejs` (DI) & Mikro ORM

lukasaric opened the pull request:
This PR refactors complete server-side structure.

**Core changes:**

- Typescript migration
- `Sequelize` 👉  [Mikro ORM](https://mikro-orm.io/) replacement
- DI `bottlejs` implementation

**Features:**

**Core:**
- [x] - `bottlejs` setup
- [x] - Remove `Sequelize` from codebase & Mikro Orm setup
- [x] - Typescript setup
- [x] - Config and environments setup
- [x] - Mail service implementation
- [x] - Add E2E tests
- [ ] - Add unit tests
- [ ] - #18 

**Major core:**
- [x] - Avoid passing container when registering dependencies. Instead, send dependencies as params which will be properly typed.

**Auth:**
- [x] - Auth implementation (basic `jwt`  and `local` strategy)

**Seeds & CLI implementation:**
- [x] - `add-user` script
- [x] - #20
- [x] - #21 

**Storage:**
- [x] - Rewrite to TS
- [x] - #19 

**Datasheet:**
- [ ] - Rewrite `datasheet` file to TS

**User module:**
- [x] - Implement CRUD users
- [x] - Add validation
- [x] - Implement `model` and `repository`
- [x] - Introduce import service within the user
- [x] - Introduce user notification service

**Extra FE adjustments:**

- [x] - Validation enhancement and fixes
- [x] - Vue CLI

lukasaric committed:
Removed non-existing ejs setup

lukasaric committed:
Storage providers cleanup

lukasaric committed:
Replace void with any on storage wrapper methods

lukasaric committed:
Bump inquirer & exceljs

lukasaric committed:
Use repo in user script to invoke encrypt

lukasaric committed:
Adjusted eslint for FE

zcuric commented on server/shared/datasheet.js:
```diff
@@ -1,9 +1,7 @@
-const intoStream = require('into-stream');
-const { Workbook } = require('exceljs');
-const Worksheet = require('exceljs/dist/es5/doc/worksheet');
+import { camelCase } from 'change-case';
```
@malizeus I think `datasheet.js` has been improved a lot in `boutique`. Maybe you could check it and implement these improvements in this stater. 👀

zcuric commented on server/shared/logger.ts:
```diff
@@ -0,0 +1,20 @@
+import Logger, { LogLevel, Stream } from 'bunyan';
```
@malizeus Have you guys considered moving away to `bunyan` to something more stable like `pino`?

lukasaric committed:
Remove old user module & update eslint & deps

lukasaric committed:
Introduce user import service

lukasaric committed:
Add password to be optional param in ctr

lukasaric committed:
Uncomment jwt authenticate :clown_face:

lukasaric committed:
Added fullName & label prop getters

lukasaric committed:
Added archived flag within the pagination

lukasaric committed:
Remove use-stricts

lukasaric committed:
fullName returning type improvement

lukasaric committed:
Fixed wrong isArchived cond

lukasaric committed:
Remove console.log :clown_face:

lukasaric committed:
Replace `find` with `findAndCount`

lukasaric committed:
Added missing import service interface

lukasaric committed:
Fixed reverse proxy port

lukasaric committed:
Fixed user list response

lukasaric committed:
Added create or restore feature

lukasaric committed:
Fixed showing archived users

lukasaric committed:
Bulk import errs check cleanup

lukasaric committed:
Remove unnecessary async

lukasaric committed:
Use context for user istead of polluting the req

lukasaric committed:
Renamed targetUser to user

zcuric commented on server/auth/middleware/authorize.ts:
```diff
@@ -1,10 +1,12 @@
 import roles from '../../user/roles';
 
-function authorize(allowed: string[]): RequestHandler {
+function authorize(allowed: string[] | string[]): RequestHandler {
```
You probably doubled this wrongly. :)

lukasaric committed:
Removed allowed arg type redundancy :clown_face:

lukasaric committed:
Registered pagination as DI mw

lukasaric committed:
Fixed env example log lvl

lukasaric committed:
Mailtrap abstraction & structure improvement

lukasaric committed:
Move parser opts to overrides scope

lukasaric committed:
Config improvements :rocket:

lukasaric committed:
Cover storage provider config validation

lukasaric committed:
Renamed Scope to AudienceScope

lukasaric committed:
Added port resolvement within the poi

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into add-vue-cli

lukasaric committed:
Enhance isDev env check

lukasaric committed:
Eslint minor source type module coverage

lukasaric committed:
Added missing NextFunction type

lukasaric committed:
Fixed server build

lukasaric committed:
Send deps instead of whole container :rocket:

lukasaric committed:
Updated hygen templates

lukasaric committed:
Removed unnecessary IContainer import

lukasaric committed:
Move invite within the import service

lukasaric committed:
Remove unused import

lukasaric committed:
Changed build output dirs for FE & BE

lukasaric committed:
Improve import interface nomenclature

lukasaric committed:
Renamed paranoid to archived :eyes:

lukasaric committed:
Environments cleanup

lukasaric committed:
Use prod & dev node env vars

lukasaric committed:
Cleanup

lukasaric committed:
Storage config & params type check improvement

MiroDojkic committed:
Fix vee unexisting rule reference  🐛

lukasaric committed:
Added missing confirmed validation rule

lukasaric committed:
Revert confirmed rule changes

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into add-vue-cli

lukasaric committed:
Minor eslint overrides files correction

lukasaric committed:
Fixed main path in pkg json

MiroDojkic committed:
Replace `is` rule with `confirmed` rule 🎨

lukasaric committed:
Merge pull request #16 from ExtensionEngine/fix/reset-password-confirm-validation

Fix: Unexisting Vee Validate rule referenced in reset password field

lukasaric committed:
Fix lint issues

lukasaric committed:
FE validation enhancement :surfer:

- Set default interactive mode to be passive
- Use object rules syntax instead of string :art:
- Use name_format validation on name fields

lukasaric committed:
Use common dialog as confirm dialog wrapper

lukasaric committed:
Upgrade deps

lukasaric committed:
Removed unnecessary asyncs

MiroDojkic committed:
Rename IEnv type to Env 🎨

MiroDojkic committed:
Move authentication provider setup to service ♻️

lukasaric committed:
Remove `I` from middleware types

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into fix/refactor-auth-initialize-middleware

lukasaric committed:
Inline ctr args

lukasaric committed:
Allow error to be null in SecretOrKeyCallback

lukasaric committed:
Cleanup :lipstick:

- Removed unnecessary & ordered imports
- Used Paramates ts util type in handle
- Removed unnecessary semicolons

lukasaric committed:
Err mw interface fix

lukasaric committed:
Add database host in db config :clown_face:

lukasaric committed:
Removed `I` from Env type

lukasaric committed:
Order :eyes:

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into fix/refactor-auth-initialize-middleware

lukasaric committed:
Set min npm version to be 6 in pkg json

lukasaric committed:
Changed dev server script name

MiroDojkic committed:
Revert middleware interfaces rename ↩️

MiroDojkic committed:
Merge pull request #17 from ExtensionEngine/fix/refactor-auth-initialize-middleware

Move authentication provider setup to AuthService

lukasaric committed:
Unified storage methods within the IStorage

lukasaric committed:
Changed write stream creation interface

lukasaric committed:
Cleanup

lukasaric committed:
Exposed createStorage fn

lukasaric committed:
Added missing semicolons

lukasaric committed:
Init storage provider in main

lukasaric committed:
Fixed exists method :wrench:

lukasaric committed:
Filesystem methods cleanup :nail_care:

lukasaric committed:
Removed mode param

lukasaric committed:
Amazon methods cleanup :art:

lukasaric committed:
Used FileListResponse type for list files

lukasaric committed:
Removed unnecessary validation

- Validation is covered when creating storage config

lukasaric committed:
Add extra provider config extraction type check

lukasaric committed:
Scripts naming :eyes:

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Replace origin resolvement with server_url env var

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Added port in env example

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Introduced Commander CLI :rocket:

lukasaric committed:
Replace exists with fs exists

lukasaric committed:
Password encryption refactor :rocket:

- Removed user subscriber
- Encrypt pw within the entity
- Removed salt rounds from env, used hc value
- Renamed password to password_hash column

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-cli-manager

lukasaric committed:
Added seed service :rocket:

lukasaric committed:
Added users seed

lukasaric committed:
Removed console.log :eyes:

lukasaric committed:
Removed unnecessary use-strict

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-cli-manager

lukasaric committed:
Merge branch 'feature/add-cli-manager' into feature/add-seeds

lukasaric committed:
added db:seed script :scroll:

lukasaric committed:
Send em as param in seed utils

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Removed unnecessary promise resolve

lukasaric committed:
Use fs exists method

lukasaric committed:
Removed unnecessary castin & path resolve fix

lukasaric committed:
Class members order :eyes:

lukasaric committed:
Init storage within the createStorage fn

lukasaric committed:
Set proper return type on write stream method

lukasaric committed:
Set proper return type for file exists method

lukasaric committed:
Delete response type & methods enhancement

lukasaric committed:
Remove unused not found msg

lukasaric committed:
Changed file list response type to have path only

lukasaric committed:
Remove Response interface

lukasaric committed:
Set raw prop to be Buffer :wrench:

lukasaric committed:
Added AWSError type check

lukasaric committed:
Add system err interface :eyes:

lukasaric committed:
Added semicolons to type definitions

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Defined responses as types

lukasaric committed:
Removed unnecessary use-strict's

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-cli-manager

lukasaric committed:
Merge branch 'feature/add-cli-manager' into feature/add-seeds

lukasaric committed:
Added types to addUser & validate fn's

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into add-vue-cli

lukasaric committed:
Fixed pkg lock

MiroDojkic committed:
Add cypress ➕

Commit:
Prepared config and data files

Commit:
Prepared config and data files

lukasaric committed:
Renamed env port to server port

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into add-vue-cli

Commit:
Added page objects and selectors, introduced xpath

lukasaric committed:
Vue config adjustments :wrench:

- Added public folder
- Added page titles
- Removed all build plugins
- Set `VUE_APP` prefix for client vars
- Deps cleanup :package:

lukasaric committed:
Added global scss variables

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Upgrade sass & sass-loader

lukasaric committed:
Encoded copy source file name

lukasaric committed:
Merge pull request #6 from ExtensionEngine/add-vue-cli

Introduce vue-cli

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-cli-manager

lukasaric committed:
Merge branch 'feature/add-cli-manager' into feature/add-seeds

lukasaric committed:
Added APP_URL env

Commit:
Debbuging some issues

Commit:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into test/cypress

Commit:
package.json

Commit:
Fixed issues, completed first test.

Commit:
Updated login test

Commit:
Updated login test

lukasaric committed:
Move SheetData type to import service interface

Commit:
Removed page objects and added logout and reset password tests

Commit:
Removed page objects and added logout and reset password tests

Commit:
Add user test

Commit:
Add user test

Commit:
disabled screenshot and video on test fail

Commit:
disabled screenshot and video on test fail

lukasaric committed:
Use base axios instance for reset pw

lukasaric committed:
Use JwtPayload interface

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

MiroDojkic committed:
Merge pull request #20 from ExtensionEngine/feature/add-cli-manager

Introduce Commander CLI manager

lukasaric committed:
Addressed issues :lipstick:

- Resolved provider path within the cr8 storage fn
- Set return type for file deletion to be Promise<void>
- Removed DeleteResponse type
- Amazon config setup in ctr adjustments
- Extracted noop fn
- Aligned writeStream creation to be resolved as promise

ikovac committed:
Add env example 🔧

ikovac committed:
Inject envs into cypress env config ✨

ikovac committed:
Add cypress login command for programmatically login ✨

ikovac committed:
Remove console logs 🔥

ikovac committed:
Add user login envs 🔧

ikovac committed:
Add login actions ✨

ikovac committed:
Delete data.json 🔥

lukasaric committed:
isNotFound error improvement

lukasaric committed:
Export providers within the index

lukasaric committed:
Remove empty line

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-seeds

ikovac committed:
Delete fixteures config 🔥

ikovac committed:
Save user to local storage after successfull login ♻️

ikovac committed:
Add app url env as baseUrl ♻️

ikovac committed:
Add app url to env example 🔧

lukasaric committed:
Separated admin & users seed

ikovac committed:
Improve admin add user test ✅

lukasaric committed:
Seed admin first

lukasaric committed:
Move seeding admin within the users module

lukasaric committed:
Put API docs above the s3 init

lukasaric committed:
Remove console.log :clown_face:

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

MiroDojkic committed:
Merge pull request #19 from ExtensionEngine/feature/storage-abstraction-enhancement

Storage abstraction enhancement

lukasaric committed:
Add missing pw within the user seed

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-seeds

lukasaric committed:
Handle seed all within the Seed service

ikovac committed:
Add reset database command ✨

ikovac committed:
Add seed action ✨

ikovac committed:
Seed database before each test ✅

lukasaric committed:
Move seed service in seeds folder

lukasaric committed:
Renamed rng migration name to CreateUser

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/add-seeds

Commit:
Added user seed to tests

Commit:
Added data-testid on the admin avatar

MiroDojkic committed:
Merge pull request #21 from ExtensionEngine/feature/add-seeds

Introduce seed service

MiroDojkic committed:
Shorten cy script aliases 🎨

MiroDojkic committed:
Use single dotenv file for both app and tests 🔧

MiroDojkic committed:
Move require above local declarations 🎨

MiroDojkic committed:
Update login/logout test descriptions 🎨

MiroDojkic committed:
Update add user test to avoid using element selectors

MiroDojkic committed:
Remove commented support file imports, trailing comma 🎨

MiroDojkic committed:
Add cypress ➕

MiroDojkic committed:
Added page objects and selectors, introduced xpath

MiroDojkic committed:
Debbuging some issues

MiroDojkic committed:
package.json

MiroDojkic committed:
Fixed issues, completed first test.

MiroDojkic committed:
Fixed tests and cleaned files

MiroDojkic committed:
Align config param names 🐛

MiroDojkic committed:
Search for user before assertion in add-user test ✅

MiroDojkic committed:
Inline single method call 🎨

MiroDojkic committed:
Replace promise chain with async await 🎨

MiroDojkic committed:
Reorder testid props 🎨

lukasaric committed:
Add missing autobind on Seed service

MiroDojkic committed:
Increase navbar avatar test ID specificity 🎨

MiroDojkic committed:
Add EOF newline to .gitignore 🚨

MiroDojkic committed:
Replace anonymous function with arrow 🎨

Commit:
Removed leftover .env.example from cypress folder

Commit:
Updated .env.example to include test configuration

Commit:
Merge branch 'test/cypress' of https://github.com/ExtensionEngine/app-starter into test/cypress

nikola-zitko committed:
Replaced all anonymous functions with arrow

lukasaric committed:
Added missing `I` prefix on mw interface template

MiroDojkic commented on server/user/interfaces/notification.service.ts:
```diff
@@ -0,0 +1,9 @@
+
+import User from '../../user/model';
+
+interface IAuthService {
```
```suggestion
interface IUserNotificationService {
```

lukasaric committed:
Fixed notification interface naming

lukasaric committed:
Removed unnecessary user repo prefix in controller

lukasaric committed:
Replaced Promise.all with P.each

MiroDojkic committed:
Rename reset-db file 🚚

MiroDojkic committed:
Update reset password test description 🎨

MiroDojkic committed:
Update scaffolder templates 🔨

MiroDojkic committed:
Rename unused variable to _ 🎨

MiroDojkic committed:
Replace map over promise array with Bluebird#each 🎨

MiroDojkic committed:
Fix UserNotificationService types and order 🎨

MiroDojkic committed:
Align user controller method names 🎨

MiroDojkic committed:
Rename user middleware files to kebab-case 🎨

MiroDojkic committed:
Add missing ; for type declaration ✏️

lukasaric committed:
Db migrations naming strategy improvement :recycle:

lukasaric committed:
Merge pull request #26 from ExtensionEngine/fix/ts-migration-leftover

TS migration fixes and adjustments

lukasaric committed:
Merge branch 'test/cypress' of github.com:ExtensionEngine/app-starter into test/cypress

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into test/cypress

lukasaric committed:
Add private instance getter withing the repo :recycle:

lukasaric committed:
Merge pull request #24 from ExtensionEngine/test/cypress

Setup Cypress and add auth and admin add user tests

lukasaric committed:
Add templates build & improved lint scripts :art:
